### PR TITLE
[Vouchers] Percentage rate

### LINKED ISSUE #10727: [Vouchers] Include unfixed amounts and percentage rates
## Description

Extending the first bare minimum iteration of vouchers this next stage will expand the functionality to include:
- Both Percentage and Flat Amount for Voucher Type
- The Amount can be set to anything. For percentage this is between 0 and 100 as a double. For Flat Amount this is a non-negative double. 
Screenshot from 2023-04-20 07-32-19
_Screenshot from designs [here](https://www.figma.com/file/ZOe1Xmz4kg9wnHw1XFu389/Voucher-Back-office?node-id=47-301)_

The rate will be reflected in the list page and will be calculated accurately at checkout.

## Acceptance Criteria & Tests

1. Create voucher codes in the backoffice - using a range of percentages to test edge cases (0, 100 etc) and amounts (more than order total, exactly order total etc)
2. Apply to a checkout flow and see the discount applied
3. If the order total goes below zero when the code is applied it should stay at zero
4. Disregard (at this stage) any case that additional negative fees apply to the order
5. Voucher codes should be able to be used unlimited number of times
6. Voucher codes should apply to any shoppers
7. Voucher codes should not apply in enterprises that have not enabled them. If the enterprise has enabled no voucher code the voucher code box should not be visible at checkout
8. If the voucher code covers the cost of the whole order, just display the full range of payment methods, no clever logic at this stage.
10. Translations as standard
11. Unit tests as standard

### PR DESCRIPTION
**Pull request by rioug**
:information_source: Please track code review under the project **`##10430 Vouchers - Aus Pt 2 (UK)`**.

#### What? Why?

- Last part of  #10727 

Add the ability to have a percentage based voucher. It supports tax included in price and tax excluded from price to same extend that flat rate voucher ie. we can't have order which mix tax included and tax exclude from price.

See calculation reference here : https://github.com/openfoodfoundation/openfoodnetwork/issues/10432#issuecomment-1439402476

#### What should we test?

* Enable vouchers feature
* Visit admin/feature-toggle/features and enable or add vouchers feature

In the backoffice logged as an enterprise user:

* Visit the Enterprise tab and click on "Settings" for a shop of your choosing, and click on the "vouchers" in left hand side menu
* Click on "Add New button" --> New Voucher page should now have a "Voucher type" and an "Amount" input
* Enter a voucher code, an amount and choose "Percentage(%)" type , and click "Save"
    --> It should redirect you to the vouchers page on the enterprise edit page
    --> You should now see a list of vouchers with the one you just created, included the entered percentage rate

On the website logged as a customer :

* Add one or more product in your cart from the shop above
* Proceed to checkout
* Enter details and shipping info, click on "Next - Payment method"
* Enter the voucher code and click "Apply"
    --> You should see a "xx % Voucher" box where xx is the amount you entered when creating a voucher
* choose a payment method and click on "Next - Order summary"
    --> you should see the voucher applied to the order in the order summary column, with the correct amount
* Click on "Complete order"
    --> you should see the voucher applied to the order in the order summary with the correct amount

Repeat the above scenario with tax included in price and tax excluded from price

##### Payment fees and voucher 
Payment fees are calculated based on the item total, and are recalculated when confirming the order. Vouchers are calculated based on the order total. The main thing to check here, is to make sure : 
* voucher calculation include the payment fee 
* when using a percentage based payment fee, the payment fee shouldn't change when confirming an order with a voucher
* when a voucher cover the order total or more, -> No payment required, so the payment fee shouldn't apply

See this https://github.com/openfoodfoundation/openfoodnetwork/pull/11117#discussion_r1251505213 for more background 

#### Release notes

Changelog Category: User facing changes

The title of the pull request will be included in the release notes.

#### Dependencies
This is base on this PR #11117, ~~which depends on #11135~~. They will need to be merged first.

### CODE REVIEW

**Comment by rioug** [comment]
New commit start here: https://github.com/openfoodfoundation/openfoodnetwork/pull/10821/commits/28b3163ede57e62b7a18ab25bb506da3f5681bd4

Admin screen and Calculation are done.
There is currently an issue with percentage voucher where the adjustment keeps getting recalculated during checkout, resulting in the wrong amount for the adjustment. As far as I can tell it's due to `Spree::Adjustment.update_adjustment!` being called each time  `@order.update_order!` is called. So If the `order.total` has been updated to take in to account the voucher adjustment, any new call to `Spree::Adjustment.update_adjustment!(calculable = nil, force: false)` will update the voucher adjustment amount, because the calculation is based on the `order.total` see `Voucher#compute_amount`
Potential fix : 
- prevent voucher adjustment from being updated in `Spree::Adjustment.update_adjustment!` 
- refactor `Voucher#compute_amount` so it doesn't calculate the voucher amount based on the order.total, and maybe move this calculation in `VoucherAdjustmentsService`


TODO:
- system spec for percentage voucher

**Comment by rioug** [comment]
From @mkllnk on slack : 

> refactor `Voucher#compute_amount` so it doesn't calculate the voucher amount based on the order.total, and maybe move this calculation in `VoucherAdjustmentsService`


This sounds like the long-term solution that also works with calculating tax adjustments for mixed in-ex-cluded taxes.

**Comment by rioug** [comment]
Reviewers new commit starts here b461dfdd921a5c8742881a470129c65c3744b314

File: app/models/voucher.rb
Comment (mkllnk) [inline app/models/voucher.rb:34]: Why not use Rails helpers? 

* https://apidock.com/rails/v5.2.3/ActionView/Helpers/NumberHelper/number_to_percentage

File: app/models/voucher.rb
Comment (mkllnk) [inline app/models/voucher.rb:64]: I think I would find `if else` easier to read here. Or a `case` statement.  It's not a guard clause. The two results are of equal status. It's opinionated and I'm probably thinking about it too much but I would have done this:


```suggestion
    total = order.pre_discount_total

    value = case voucher_type
            when FLAT_RATE
              amount
            when PERCENTAGE_RATE
              amount / 100 * total
            else
              0
            end

    value.clamp(0, total) * -1
```

Many more lines but I find it clearer.

**Review by mkllnk — COMMENTED** [review]
WIP - didn't finish the review. I'll continue tomorrow.

File: app/services/voucher_adjustments_service.rb
Comment (mkllnk) [inline app/services/voucher_adjustments_service.rb:36]: Maybe this should be a method on the voucher so that it can be re-used.

File: spec/system/consumer/split_checkout_tax_not_incl_spec.rb
Comment (mkllnk) [inline spec/system/consumer/split_checkout_tax_not_incl_spec.rb:156]: Do we need these comments?

**Review by mkllnk — APPROVED** [review]
Nice work. I have a few comments but nothing blocking this.

File: db/migrate/20230508035306_add_type_to_vouchers.rb
Comment (rioug) [reply db/migrate/20230508035306_add_type_to_vouchers.rb:3]: This seems the perfect use case for it, I'll give it a go.

Comment (mkllnk) [reply db/migrate/20230508035306_add_type_to_vouchers.rb:3]: The other option would be to introduce a polymorphic model. You create two models for percentage and flat amount and introduce the magic Rails `type` column.

Comment (rioug) [reply db/migrate/20230508035306_add_type_to_vouchers.rb]: Ended up using a single table inheritance with the `type` collumn

File: spec/system/consumer/split_checkout_tax_not_incl_spec.rb
Comment (rioug) [reply spec/system/consumer/split_checkout_tax_not_incl_spec.rb]: Removed !

**Comment by filipefurtad0** [comment]
Hey, there are some conflicts blocking testing on this one. Ping @rioug.

**Comment by rioug** [comment]
New commit starts here https://github.com/openfoodfoundation/openfoodnetwork/pull/10821/commits/cb54397263662c5ba5cea23f2fb28fe07d772765

File: app/models/vouchers/percentage_rate.rb
Comment (mkllnk) [inline app/models/vouchers/percentage_rate.rb]: Was the comment meant for the clamping of the flat rate voucher?

We limit the rate to a maximum of 100% in the percentage case which can't exceed the total by definition.

File: app/models/vouchers/percentage_rate.rb
Comment (mkllnk) [inline app/models/vouchers/percentage_rate.rb]: I wonder if we should add `compute_tax_amount` to the model as well... or take this out of the model. But then this is to comply with the originator interface of the Spree adjustments, right? Nevermind...

File: app/controllers/admin/vouchers_controller.rb
Comment (mkllnk) [inline app/controllers/admin/vouchers_controller.rb:22]: You could `constantize` the type string after checking it against the known types. Maybe worth if we add more types.

File: app/controllers/admin/vouchers_controller.rb
Comment (mkllnk) [inline app/controllers/admin/vouchers_controller.rb:25]: I would love to phase out the use of "Spree". This is not a Spree controller.

File: db/migrate/20230508035306_add_type_to_vouchers.rb
Comment (mkllnk) [inline db/migrate/20230508035306_add_type_to_vouchers.rb:6]: Sometimes I wonder if it's more efficient to say `default: "Vouchers::FlatRate"` so that the database initialises the value. Uh, that reminds me: should this be `null: false`?

**Review by mkllnk — APPROVED** [review]
Nice refactoring!

I have some comments but nothing blocking the delivery of this part.

File: app/models/vouchers/percentage_rate.rb
Comment (rioug) [reply app/models/vouchers/percentage_rate.rb]: yeah it was, I'll clean that up

File: app/models/vouchers/percentage_rate.rb
Comment (rioug) [reply app/models/vouchers/percentage_rate.rb]: Yes the `compute_amount` is there to comply with the originator interface, see `CalculatedAdjustments` concern. That said, I believe voucher adjustments are decoupled from other adjustment lifecycle, so I think we "should be able" to take it out of the model.

File: app/controllers/admin/vouchers_controller.rb
Comment (rioug) [reply app/controllers/admin/vouchers_controller.rb:22]: something like ?  
```ruby
voucher_types = ["Vouchers::FlatRate", "Vouchers::PercentageRate"]
if voucher_types.include?(params[:voucher][:voucher_type])
  @voucher = params[:voucher][:voucher_type].constantize.create(permitted_resource_params.merge(enterprise: @enterprise))
else 
  @voucher.errors.add(:type)
  return render_error
end
```

File: app/controllers/admin/vouchers_controller.rb
Comment (rioug) [reply app/controllers/admin/vouchers_controller.rb]: Done, I saw that this translation also exists at the root level

File: db/migrate/20230508035306_add_type_to_vouchers.rb
Comment (rioug) [reply db/migrate/20230508035306_add_type_to_vouchers.rb]: Nice one, indeed null should be false. I meant the code guards against it has it would blow up :laughing: but better to enforce it at the database level.

File: app/controllers/admin/vouchers_controller.rb
Comment (rioug) [reply app/controllers/admin/vouchers_controller.rb:14]: I think this is safe to ignore, as we are doing whitelist check before calling `safe_constantize`, it's a known brakeman false positing issue : https://github.com/presidentbeef/brakeman/issues/1115

**Review by dacook — APPROVED** [review]
Great, looks good to me!

I have a few comments, but I think they're more just personal preference and don't need a change.

File: spec/system/consumer/split_checkout_tax_not_incl_spec.rb
Comment (dacook) [reply spec/system/consumer/split_checkout_tax_not_incl_spec.rb:206]: I just realised that we have this setting enabled, so there shouldn't be any linebreaks getting in the way: 
https://github.com/openfoodfoundation/openfoodnetwork/blob/master/spec/system/support/capybara_setup.rb#L8-L11

File: spec/models/vouchers/flat_rate_spec.rb
Comment (rioug) [reply spec/models/vouchers/flat_rate_spec.rb]: They don't indeed, I think it's a left over from when there wasn't a voucher factory. I clean up the specs

File: app/controllers/admin/vouchers_controller.rb
Comment (rioug) [reply app/controllers/admin/vouchers_controller.rb:16]: I added a comment in the code, seems like the easiest solution. I guess we can look into adding `brakeman.ignore` file if it becomes problematic

File: app/controllers/admin/vouchers_controller.rb
Comment (dacook) [inline app/controllers/admin/vouchers_controller.rb]: If it is `vouchers_flat_rate`, then would the `voucher_type` always be flat rate?

**Review by dacook — APPROVED** [review]
👍

File: app/controllers/admin/vouchers_controller.rb
Comment (dacook) [reply app/controllers/admin/vouchers_controller.rb]: I just randomly came back to this and  have to admit confusion. 

If the form is rendered with 
```
    def new
      @voucher = Voucher.new
    end
```
and
```
= form_with model: @voucher, ...
```

Then why are the form fields named after one of the sub-classes `vouchers_flat_rate`?

It seems to work according to the spec so am I missing something?

Comment (rioug) [reply app/controllers/admin/vouchers_controller.rb]: It is a bit confusing indeed, it is because `Vouchers::FlatRate` is the default type on `Voucher#type`, see migration below: 
```ruby
  def up
    # It will set all the vouchers till now to Vouchers::FlatRate
    add_column :vouchers, :type, :string, limit: 255, null: false, default: "Vouchers::FlatRate"
  end
```

So when you do ` @voucher = Voucher.new` you actually create a `Vouchers::FlatRate` hence why the form filed are name `vouchers_flat_rate`

Comment (dacook) [reply app/controllers/admin/vouchers_controller.rb]: Ah ha, thanks for pointing that out. I do remember seeing that last week now..

File: app/controllers/admin/vouchers_controller.rb
Comment (dacook) [reply app/controllers/admin/vouchers_controller.rb]: @mkllnk I think this is just waiting on your re-review of the fix for this comment (see newer commit https://github.com/openfoodfoundation/openfoodnetwork/pull/10821/commits/229ae91f3f947e825386d1d6063bbcb73ebbd700)

File: app/controllers/admin/vouchers_controller.rb
Comment (mkllnk) [reply app/controllers/admin/vouchers_controller.rb:16]: That's unfortunate. Another solution would be select *our* string based on the params value so that we are not using user input directly. But it's really the same.

**Review by mkllnk — APPROVED** [review]
All good.

If you request a review from me via Github, I may realise that earlier and respond. :smile:

**Comment by drummer83** [comment]
Hi @rioug,
Thanks for this feature. Taxes are horrible, I hope I am getting this right. :crossed_fingers: 

# Summary
## Questions (for product)
- Editing the order in the back office leaves the payment fees and voucher discount untouched. Is this correct behavior? One could argue that a percentage voucher should apply to added products as well. :question: 
- Should the display of the tax amount for tax excluded from price be combined in one final tax amount instead of displaying the original tax amount and the discount? It's irritating and difficult to understand. See **DETAIL 1** below. :question: 
- Should it be possible to create vouchers with 0 % amount? Current minimum is 0.01 %. :question: 
- Do we need to improve the edge case around discounts of 0.01 $? See **DETAIL 2** below. :question: 

## Errors (needs fixing)
- Error 500 when trying to create a percentage voucher after trying a first time without entering a voucher code. See **DETAIL 3** below. :x:
- For tax included in price the tax amount is not updated after applying the voucher. See **DETAIL 4** below. :x:
- Taxes in reports are not correct, e.g. Sales Tax Totals By Order report. See **DETAIL 5** below. :x:

## Improvements
- The note 'Note: if your order total is less than your voucher you may not be able to spend the remaining value.' shouldn't be displayed for percentage vouchers because in this case it can't happen.

## Conclusion
If you agree let's merge this one and open issues for the remaining errors and other bits which may come up after the questions have been answered.
What do you think?

# Details to the above
## Test scenarios (gathered from issue, discussions and other related PRs)
- Vouchers with percentage can be created. :heavy_check_mark: 
- Vouchers with percentage are listed on the vouchers page. :heavy_check_mark: 
- Payment fees are calculated based on line items, and don't change with applied vouchers. Discount is calculated on line items + payment fees. :heavy_check_mark: 
- Vouchers discount the order total after applying tax. Tax adjustments are listed as order adjustments. 
  - **DETAIL 4:** If tax is included in price there are two adjustments needed (see first example [here](https://github.com/openfoodfoundation/openfoodnetwork/issues/10432#issuecomment-1439402476)):
    - Reduce the order total (amount). :heavy_check_mark: 
    - Reduce the included tax. This is not happening currently, the tax included value is still calculated on the order total without the discount. :x:
image
image
  - **DETAIL 1:** If tax is excl. from price we have two adjustments as described [here](https://github.com/openfoodfoundation/openfoodnetwork/issues/10432#issuecomment-1439402476) as well. The values are correct but I find it very misleading to have the additional adjustment on tax. The customer can't know if the 4.00 $ in the example below is already the discounted tax or the original tax. He has to calculate himself. Why don't we do that for him by only displaying the final tax paid? I'm not sure if it's technically feasible but we could first apply the discount and then calculate the tax or simply add the original tax and the negative tax discount and display only this result. :question: 
image

### Acceptance criteria from issue

1. Create voucher codes in the back office - using a range of percentages to test edge cases (0, 100 etc) and amounts (more than order total, exactly order total etc)
- Can't create voucher with 0 % discount. Amount has to be >= 0.01 %. :question: 
- **DETAIL 2:** The edge case of discounts around 0.01 $ is not covered properly. Example: Order total = 0.49 $ without tax; Added tax = 0.05 $; Voucher amount = 1 %; 1 % of 0.49 $ is less than 0.01 $ and will not be displayed. 1 % of 0.05 $ is less than 0.01 $ and will not be displayed. --> The voucher is not applied. BUT: The Order total including tax is 0.54 $. 1 % of 0.54 $ is 0.0054 $ which is rounded to 0.01 $ so the customer should get a discount of 0.01 $ . I don't think a case like this one is likely to happen though. :question: 
2. Apply to a checkout flow and see the discount applied :heavy_check_mark: 
3. If the order total goes below zero when the code is applied it should stay at zero
This is not possible using a percentage rate voucher. 
4. Disregard (at this stage) any case that additional negative fees apply to the order
5. Voucher codes should be able to be used unlimited number of times :heavy_check_mark: 
6. Voucher codes should apply to any shoppers :heavy_check_mark: 
7. Voucher codes should not apply in enterprises that have not enabled them. If the enterprise has enabled no voucher code the voucher code box should not be visible at checkout :heavy_check_mark: 
8. If the voucher code covers the cost of the whole order, just display the full range of payment methods, no clever logic at this stage.
There used to be a 'No payment required' note, which currently is not displayed during first checkout anymore. Not even with a 100 % voucher applied. It is only displayed when moving back to step 2 after being on step 3 already (see [comment in #11117](https://github.com/openfoodfoundation/openfoodnetwork/pull/11117#issuecomment-1649603057)).
9. Translations as standard :heavy_check_mark:  

- Flat amount vouchers still work? :heavy_check_mark: 
- Mix with no tax works?  :heavy_check_mark: 
- **DETAIL 5:** Are taxes correct in the tax report(s)? :x:
I checked the Sales Tax Totals By Order report. Other reports may be affected as well. The values there are not correct:
  - Discounts on taxes should be reflected in the tax column (= original tax minus tax discount).
  - Total without tax needs to be updated accordingly.
image

- Are all types of fees working correctly? :question: 
I don't know. Let's test that at a later point in time.

### Errors
- **DETAIL 3:** Error 500 when trying to create a percentage voucher after trying a first time without entering a voucher code. [Bugsnag](https://app.bugsnag.com/yaycode/openfoodnetwork-aus/errors/64c25864c4516800087f95dd?filters[event.since]=30d&filters[error.status]=open&filters[app.release_stage]=staging) Same happens if the voucher amount was not a number or zero or code taken or amount >100  on the first try. For flat amounts this is working ok.

https://github.com/openfoodfoundation/openfoodnetwork/assets/1790176/c5a03107-23fc-4d89-bb8e-f0f5311e2e24

Thanks again!!

**Comment by rioug** [comment]
> For tax included in price the tax amount is not updated after applying the voucher. See DETAIL 4 below. x

In this scenario, the the tax amount is included in the voucher adjustment, but it's not displayed. In your scenario the included tax would be $0.09 -> Included tax $0.91- Voucher included tax $0.09 = $0.82. From memory the design didn't show the voucher included tax, so I am not sure if this need fixing ? 

> Taxes in reports are not correct, e.g. Sales Tax Totals By Order report. See DETAIL 5 below. 

I agree we should fix the report, but I am not sure this is in scope ?

**Comment by lin-d-hop** [comment]
Hey team, Great testing @drummer83!

_1) Editing the order in the back office leaves the payment fees and voucher discount untouched. Is this correct behavior? One could argue that a percentage voucher should apply to added products as well. question_

I am happy with this the way it is. It could go either way and both make rational sense so let's not open this can of worms. If someone wants it in the future we can explore it then. Workaround to add an adjustment in the BO.

_2) Should the display of the tax amount for tax excluded from price be combined in one final tax amount instead of displaying the original tax amount and the discount? It's irritating and difficult to understand. See DETAIL 1 below. question_

This sounds more correct. The instances that have paid for vouchers have included tax. If this is irritating enough for excluded tax instances then we can explore a funded feature to improve for them.

_3) Should it be possible to create vouchers with 0 % amount? Current minimum is 0.01 %. question_

What's the use case for this?

_4) Do we need to improve the edge case around discounts of 0.01 $? See DETAIL 2 below. question_

We already have rounding issues in OFN. It might cause issues for some accounting at some point. But to be honest I'd rather wait and see how much people are annoyed by it before going further over budget to resolve it.

_Taxes in reports_

Unless it is vouchers that have messed up taxes in reports I agree this is out of scope. It would be great to do more work tidying up reports, but for now let's just stick to creating issues when we find things.

**Comment by drummer83** [comment]
> _3) Should it be possible to create vouchers with 0 % amount? Current minimum is 0.01 %. question_
> 
> What's the use case for this?

@lin-d-hop I don't know. It's the first of the acceptance criteria though ([link](https://github.com/openfoodfoundation/openfoodnetwork/issues/10727#issue-1676103128)):
> Create voucher codes in the backoffice - using a range of percentages to test edge cases (0, 100 etc) and amounts (more than order total, exactly order total etc)

**Comment by drummer83** [comment]
@rioug Should we merge this one and address remaining issues in separate PRs? I am happy to create the issues.
This one is big and time consuming for testing already.

**Comment by rioug** [comment]
We need to wait on this https://github.com/openfoodfoundation/openfoodnetwork/pull/11117 before merging, but yes we can create issues for the remaining problems. It'd be good to at least fix the error 500.

**Comment by rioug** [comment]
I have a fix for the 500 error but I kept it on a separate branch, I think this is big enough as it is :smile:

**Comment by drummer83** [comment]
@rioug I have created the following issues as an outcome of this testing a discussion:
#11362 
#11363
#11364
[wishlist #456](https://github.com/openfoodfoundation/wishlist/issues/456)

**Comment by drummer83** [comment]
Hi @rioug,
#11117 has been merged, so I guess we can move on and work on the issues I have opened (see post above).
Shall we resolve the conflicts and merge this one here?
Thanks!

**Comment by rioug** [comment]
Done ! Thanks for creating the issues @drummer83 !

**Pull request merged**
